src/tribler-common/tribler_common/notifier.py

```diff
@@ -0,0 +1,167 @@
+        with self.lock:
+            self.interceptors.pop(observer, None)
+        self.logger.debug(f"Remove interceptor {observer!r}")
+
+    def __getitem__(self, topic: FuncT) -> FuncT:
+        def wrapper(*args, **kwargs):
+            self.notify(topic, *args, **kwargs)
+        return cast(FuncT, wrapper)
+
+    def notify_by_topic_name(self, topic_name: str, *args, **kwargs):
+        with self.lock:
+            topic = self.topics_by_name.get(topic_name)
+        if topic is None:
+            self.logger.warning(f'Topic with name {topic_name} not found')
+        else:
+            self.notify(topic, *args, **kwargs)
+
+    def notify(self, topic: Callable, *args, **kwargs):
+        """ Notify all observers about the topic.
+
+        Сan be called from any thread. Observers will be called from the reactor thread during the next iteration
+        of the event loop.  An exception when an observer is invoked will not affect other observers.
+        """
+        self.logger.debug(f"Notification for topic {topic.__name__}")
+
+        topic(*args, **kwargs)
+
+        with self.lock:
+            interceptors = list(self.interceptors)
+            generic_observers = list(self.generic_observers)
+            observers = list(self.topics[topic])
+
+        suppress = False
+        for interceptor in interceptors:
+            if interceptor(topic, *args, **kwargs):
+                suppress = True
+        if suppress:
+            return
+
+        for observer in generic_observers:
+            # The topic is passed twice, the second topic is passed to the generic observer as the first argument
+            self._notify_threadsafe(topic, observer, topic, *args, **kwargs)  # pylint: disable=arguments-out-of-order
+
+        for observer in observers:
+            self._notify_threadsafe(topic, observer, *args, **kwargs)
+
+    def _notify_threadsafe(self, topic: Callable, observer: Callable, *args, **kwargs):
+        if not self.loop:
+            self._notify(topic, observer, args, kwargs)
+        else:
+            try:
+                self.loop.call_soon_threadsafe(self._notify, topic, observer, args, kwargs)
+            except RuntimeError as e:
+                # Raises RuntimeError if called on a loop that’s been closed.
+                # This can happen on a secondary thread when the main application is shutting down.
+                # https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_soon_threadsafe
+                self.logger.warning(e)
+
+    def _notify(self, topic: Callable, observer: Callable, args: tuple, kwargs: dict):
+        self.logger.debug(f"Calling observer {observer!r} for topic {topic.__name__}")
+        try:
+            observer(*args, **kwargs)
+        except Exception as e:  # pylint: disable=broad-except
+            self.logger.exception(e)
+
+    @contextmanager
+    def capture_notifications(self, *topics, suppress=True):
```

Comment:
**Suggestion:** As far as I can see, the `capture_notifications` method is used to streamline the testing process. Usually, I'm not a big fan of having code in the core that's only there to support tests. If `capture_notifications` is indeed only used for tests, should it be moved to the tests code base?

Reply:
You are right; I removed unnecessary methods and simplified the notifier logic.